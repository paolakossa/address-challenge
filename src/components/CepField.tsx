import React, { useReducer, useState } from "react";
import { useController, useFormContext } from "react-hook-form";
import Cep from "@/models/Cep";
import { MagnifyingGlassIcon } from "@heroicons/react/24/outline";

type CepCacheState = {
  [cep: string]: Cep;
};

type CacheAction = { type: "ADD_CEP"; cep: string; data: Cep };

function cacheReducer(
  state: CepCacheState,
  action: CacheAction
): CepCacheState {
  switch (action.type) {
    case "ADD_CEP":
      const newState = { ...state, [action.cep]: action.data };
      localStorage.setItem("cepCache", JSON.stringify(newState));
      return newState;
    default:
      return state;
  }
}

type CepFieldProps = {
  name: string;
  label?: string;
  valueDefault?: string;
  onCepSearch?: (cepData: Cep | null) => void;
};

const CepField = ({
  name,
  label,
  valueDefault,
  onCepSearch,
}: CepFieldProps) => {
  const { control } = useFormContext();
  const {
    field: { onChange, onBlur, value, ref },
  } = useController({
    name,
    control,
    defaultValue: valueDefault || "",
  });

  const [cacheState, dispatch] = useReducer(cacheReducer, {}, () => {
    if (typeof window !== "undefined") {
      const localData = localStorage.getItem("cepCache");
      return localData ? JSON.parse(localData) : {};
    }
    return {};
  });

  const [loading, setLoading] = useState(false);
  const [error, setError] = useState("");
  const [skipOnBlur, setSkipOnBlur] = useState(false);

  const handleCepSearch = async () => {
    const cep = value.replace(/\D/g, "");
    if (cep.length !== 8) {
      setError("CEP deve conter 8 dígitos.");
      onCepSearch?.(null);
      return;
    }

    if (cacheState[cep]) {
      setError("");
      onCepSearch?.(cacheState[cep]);
      return;
    }

    setLoading(true);
    setError("");
    try {
      const response = await fetch(`https://viacep.com.br/ws/${cep}/json/`);
      const data: Cep = await response.json();
      if (data.erro) {
        setError("CEP inválido.");
        onCepSearch?.(null);
      } else {
        dispatch({ type: "ADD_CEP", cep, data });
        onCepSearch?.(data);
      }
    } catch (err) {
      console.error("Erro na requisição do CEP:", err);
      setError("Erro ao buscar o CEP.");
      onCepSearch?.(null);
    } finally {
      setLoading(false);
    }
  };

  const handleBlur = async () => {
    if (skipOnBlur) {
      onBlur();
      setSkipOnBlur(false);
      return;
    }
    await handleCepSearch();
    onBlur();
  };

  return (
    <div>
      {label && (
        <label htmlFor={name} className="block mb-1 font-medium">
          {label}
        </label>
      )}
      <div className="flex items-center">
        <input
          id={name}
          ref={ref}
          value={value}
          onChange={onChange}
          onBlur={handleBlur}
          placeholder="Digite seu CEP"
          className="border p-2 rounded flex-1"
        />
        <button
          type="button"
          onMouseDown={() => setSkipOnBlur(true)}
          onClick={handleCepSearch}
          className="ml-2 px-4 py-2 bg-[#225741] text-white rounded cursor-pointer  hover:bg-green-700 transition"
        >
          <div className="flex items-center">
            <MagnifyingGlassIcon className="w-6 h-6" />
            <span className="mr-2">Pesquisar</span>
          </div>
        </button>
      </div>
      {loading && <p className="text-sm text-gray-500 mt-2">Buscando...</p>}
      {error && <p className="text-sm text-red-500 mt-2">{error}</p>}
    </div>
  );
};

export default CepField;
