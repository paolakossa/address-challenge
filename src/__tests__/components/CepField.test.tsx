import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import CepField from "@/components/CepField";

const Wrapper = ({ children }: { children: React.ReactNode }) => {
  const methods = useForm();
  return <FormProvider {...methods}>{children}</FormProvider>;
};

describe("CepField", () => {
  it("should render correctly with label", () => {
    render(
      <Wrapper>
        <CepField name="cep" label="CEP" />
      </Wrapper>
    );

    expect(screen.getByLabelText("CEP")).toBeInTheDocument();
  });

  it("must allow typing in the input", () => {
    render(
      <Wrapper>
        <CepField name="cep" />
      </Wrapper>
    );

    const input = screen.getByPlaceholderText("Digite seu CEP");
    fireEvent.change(input, { target: { value: "74070010" } });

    expect(input).toHaveValue("74070010");
  });

  it("should call the API when clicking the Search button", async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () =>
          Promise.resolve({
            cep: "74070-010",
            logradouro: "Avenida Independência",
            bairro: "Setor Aeroporto",
            localidade: "Goiânia",
            uf: "GO",
          }),
      })
    ) as jest.Mock;

    const onCepSearchMock = jest.fn();

    render(
      <Wrapper>
        <CepField name="cep" onCepSearch={onCepSearchMock} />
      </Wrapper>
    );

    const input = screen.getByPlaceholderText("Digite seu CEP");
    const button = screen.getByRole("button", { name: /Pesquisar/i });

    fireEvent.change(input, { target: { value: "74070010" } });
    fireEvent.click(button);

    await waitFor(() => expect(global.fetch).toHaveBeenCalledTimes(1));

    expect(onCepSearchMock).toHaveBeenCalledWith(
      expect.objectContaining({
        cep: "74070-010",
        logradouro: "Avenida Independência",
      })
    );
  });

  it("should display error when entering an invalid zip code", async () => {
    render(
      <Wrapper>
        <CepField name="cep" />
      </Wrapper>
    );

    const input = screen.getByPlaceholderText("Digite seu CEP");
    fireEvent.change(input, { target: { value: "123" } });
    fireEvent.blur(input);

    await waitFor(() =>
      expect(screen.getByText("CEP deve conter 8 dígitos.")).toBeInTheDocument()
    );
  });

  it("must use the cache when searching for a previously queried zip code", async () => {
    const onCepSearchMock = jest.fn();
    localStorage.setItem(
      "cepCache",
      JSON.stringify({
        "74070010": {
          cep: "74070-010",
          logradouro: "Avenida Independência",
          bairro: "Setor Aeroporto",
          localidade: "Goiânia",
          uf: "GO",
        },
      })
    );

    render(
      <Wrapper>
        <CepField name="cep" onCepSearch={onCepSearchMock} />
      </Wrapper>
    );

    const input = screen.getByPlaceholderText("Digite seu CEP");
    const button = screen.getByRole("button", { name: /Pesquisar/i });

    fireEvent.change(input, { target: { value: "74070010" } });
    fireEvent.click(button);

    await waitFor(() => expect(onCepSearchMock).toHaveBeenCalledTimes(1));

    expect(onCepSearchMock).toHaveBeenCalledWith(
      expect.objectContaining({
        cep: "74070-010",
        logradouro: "Avenida Independência",
      })
    );
  });

  it("must match the snapshot", () => {
    const { asFragment } = render(
      <Wrapper>
        <CepField name="cep" label="CEP" />
      </Wrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
