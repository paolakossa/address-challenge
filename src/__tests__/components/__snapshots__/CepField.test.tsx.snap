// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CepField must match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <label
      class="block mb-1 font-medium"
      for="cep"
    >
      CEP
    </label>
    <div
      class="flex items-center"
    >
      <input
        class="border p-2 rounded flex-1"
        id="cep"
        placeholder="Digite seu CEP"
        value=""
      />
      <button
        class="ml-2 px-4 py-2 bg-[#225741] text-white rounded cursor-pointer  hover:bg-green-700 transition"
        type="button"
      >
        <div
          class="flex items-center"
        >
          <svg
            aria-hidden="true"
            class="w-6 h-6"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m21 21-5.197-5.197m0 0A7.5 7.5 0 1 0 5.196 5.196a7.5 7.5 0 0 0 10.607 10.607Z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          <span
            class="mr-2"
          >
            Pesquisar
          </span>
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
